### PR TITLE
split out merch and staff merch

### PR DESCRIPTION
**Pull request by EliAndrewC**
This implements everything in #3052 
- There are now two buttons on the Merch page, one for checking for Merch and the other for checking for Staff Merch and the "got merch" states are tracked separately for each.
- This behavior is configurable so other events with a single merch location can turn it off and revert to the original behavior of all merch being combined.
- These two merch sections now reflect the staffer preference given for 2 staff shirts vs 1 event shirt.
- Donation tiers can now be expanded into their component items with a new optional config section called ``[donation_tier_items]``.  (We already have a ``[donation_tier_descriptions]`` section, but it doesn't quite do what we need.)

I still need to make additional PRs in the ``ubersystem-puppet`` and ``production-config`` repos to add support for the new ``[donation_tier_items]`` section.  Until that time, the last item above won't actually do anything for our deployments since we won't have the necessary option configured.  I should be bale to do that on Monday.

### CODE REVIEW

File: uber/models/attendee.py [hunk 1]
```diff
@@ -762,33 +764,53 @@ def merch(self):
...
+
+        if self.extra_merch:
+            merch.append(self.extra_merch)
+
+        return comma_and(merch)
+
+    @property
+    def staff_merch_items(self):
+        """Used by the merch and staff_merch properties for staff swag items."""
+        merch = []
```
Comment (EliAndrewC) [inline uber/models/attendee.py]: This method looks weird as displayed in this diff, since it has code moved here from the ``merch`` property above.  If it's difficult to follow then I recommend opening the file and just reviewing the ``staff_merch_items`` method standalone.

File: uber/templates/registration/merch.html [hunk 2]
```diff
@@ -2,6 +2,10 @@
 {% block title %}Merch Booth{% endblock %}
 {% block content %}
 
+<style type="text/css">
+    #give .merch-list { color: red; }
+</style>
```
Comment (EliAndrewC) [inline uber/templates/registration/merch.html]: Minor unrelated fix; there was a ``color: red`` style applied to the entire section where the dynamic form was created, which caused even things like button text to be red, which worked but looked weird.  This fixes that so that only the part we want to be red is red.

File: uber/tests/models/test_merch.py [hunk 3]
```diff
@@ -141,6 +141,7 @@ def test_shirt_info_marked_after_deadline(self, monkeypatch):
 class TestMerch:
     @pytest.fixture(autouse=True)
     def defaults(self, monkeypatch):
+        monkeypatch.setattr(c, 'SEPARATE_STAFF_MERCH', True)
```
Comment (EliAndrewC) [inline uber/tests/models/test_merch.py]: Ideally I'd take the time to write a separate set of tests for the case where ``SEPARATE_STAFF_MERCH`` was turned off to ensure we don't break that functionality for other events.  That didn't seem worth the time here, especially since I suspect a lot of the merch stuff may get redone down the road, but I can go in and add that if anyone thinks it's necessary.

File: alembic/versions/23193b41cfea_added_got_staff_merch.py [hunk 4]
```diff
@@ -34,7 +34,11 @@ def sqlite_column_reflect_listener(inspector, table, column_info):
 
 
 def upgrade():
-    op.add_column('attendee', sa.Column('got_staff_merch', sa.Boolean(), server_default='False', nullable=False))
+    if is_sqlite:
+        with op.batch_alter_table('attendee', reflect_kwargs=sqlite_reflect_kwargs) as batch_op:
+            batch_op.add_column(sa.Column('got_staff_merch', sa.Boolean(), server_default='False', nullable=False))
+    else:
+        op.add_column('attendee', sa.Column('got_staff_merch', sa.Boolean(), server_default='False', nullable=False))
```
Comment (EliAndrewC) [inline alembic/versions/23193b41cfea_added_got_staff_merch.py]: Whoops - I totally forgot to do the sqlite checking last time I did migrations as well.  Sorry about that - I swear I'll remember next time!

**Comment by RobRuana** [comment]
# Before: Bundle name looks like an item
screen shot 2018-01-01 at 2 54 59 pm

# After: Bundle name differentiated from items
screen shot 2018-01-01 at 3 04 36 pm

**Comment by RobRuana** [comment]
One more thing, (I think I failed to post this earlier), the reg page needs to be updated to reflect the staff shirt changes: https://github.com/magfest/ubersystem/issues/3052#issuecomment-354673471

**Comment by RobRuana** [comment]
FYI, the `pep8` errors in `groups.py` have been fixed in `master`.

**Comment by EliAndrewC** [comment]
> FYI, the pep8 errors in groups.py have been fixed in master.

Cool, I'll pull and merge master into this branch when I make the other changes.

> I get the need for donation_tier_items, but I don't really like it. I feel like we're repeating ourselves. Like, could just add an extra field to donation_tier_descriptions to hold that info? Something like items? And if items doesn't exist, we fall back to description.split('|')?

Good idea.  To be honest, I didn't notice that ``donation_tier_descriptions`` existed until after I was done.  But this should be a simple enough change regardless, so I'll go ahead and do that tonight after work.

> Also, can we change the styling to differentiate the bundle from the individual items?

The code that does the existing styling is already pretty kludgy and this will be slightly more kludgy but seems easy enough.  As long as you're okay with me implementing that in a way that doesn't fundamentally change how we're passing the data from the server to the client (which we should probably do eventually), then I'll go ahead and make that change as well after work.

**Comment by RobRuana** [comment]
The function just returns a string that we do a bunch of pattern matching on? That seems terrible. Vaya con Dios.

Also noticed this bit of code that could be be fixed in `check_merch()`:

```python
    @ajax
    def check_merch(self, session, badge_num):
        id = shirt = None
        if not (badge_num.isdigit() and 0 < int(badge_num) < 99999):
            message = 'Invalid badge number'
        else:
            results = session.query(Attendee).filter_by(badge_num=badge_num)
            if results.count() != 1:
                message = 'No attendee has badge number {}'.format(badge_num)
            else:
                attendee = results.one()
```

We're doing `results = session.query(Attendee).filter_by(badge_num=badge_num)`, which makes results a `Query` object. Then we do `results.count()`, which does a database query. Then we do `results.one()` which does _another_ database query.

We could save ourselves a query by just doing:
```python
attendee = session.query(Attendee).filter_by(badge_num=badge_num).first()
if not attendee:
    message = 'No attendee has badge number {}'.format(badge_num)
else:
   # ...
```

**Comment by EliAndrewC** [comment]
Got home from work super late tonight and didn't manage to get to this but I will for sure tomorrow.  Sorry for the delay, this is cutting much closer to the start of the event than I planned, but I'll make sure this is ready to go in time.

**Comment by RobRuana** [comment]
Don't murder yourself! If you want, I can make those suggested changes for you.

**Comment by EliAndrewC** [comment]
> The function just returns a string that we do a bunch of pattern matching on? That seems terrible. Vaya con Dios.

Yeah, it definitely would have been cleaner to do a rewrite once I implemented donation tiers.  This code predates the merch being a big list like this, so I was able to get away without any kind of structured list for a long time, and then once we needed that it was only like 4 lines of code to "just make it work" as you see.

This was a good chance to clean it up; we now return a ``merch_items`` field in addition to the textual message which the frontend can format intelligently.

> Also noticed this bit of code that could be be fixed in check_merch() ...

Good catch.  Fixed!

> the reg page needs to be updated to reflect the staff shirt changes

Done!

File: uber/configspec.ini [hunk 5]
```diff
@@ -583,6 +589,51 @@ can_volunteer   = boolean(default=True)
 consent_form    = boolean(default=False)
 
 
+[donation_tier_descriptions]
```
Comment (EliAndrewC) [inline uber/configspec.ini]: ``[donation_tier_descriptions]`` wasn't actually documented, so I added this as an explicit config section with some extensive comments for posterity.

One nice thing about taking Rob's suggestion about just adding a new item here is that I won't have to actually make a change to the puppet manifest; instead I can simply add a new sub-item to the ``donation_tier_items`` hash.

**Comment by RobRuana** [comment]
I am headed to bed, so I can’t give this a solid review. From a cursory glance, it looks great! I trust you, so feel free to merge this :)

Otherwise I’ll give it a full once-over tomorrow.

Thank you for all your hard work, on this and everything else! You never cease to amaze me :)

File: uber/models/attendee.py [hunk 6]
```diff
@@ -758,40 +759,90 @@ def donation_swag(self):
         return donation_items + extra_donations
 
     @property
-    def merch(self):
+    def merch_items(self):
```
Comment (EliAndrewC) [inline uber/models/attendee.py]: We now have both the existing ``Attendee.merch`` property which is the same as before and a new ``Attendee.merch_items`` property which returns a list of those items.  This makes it easier for the backend to pass something to the frontend for nice formatting.

File: uber/templates/registration/merch.html [hunk 7]
```diff
@@ -4,13 +4,41 @@
...
     var $shirtOpts = $('<select/>').attr('id', 'shirt');
     $.each({{ c.MERCH_SHIRT_OPTS|jsonize }}, function(i,size) {
         $shirtOpts.append($('<option/>').val(size[0]).text(size[1]));
     });
 
+    /**
+     * Given the list of items returned by Attendee.merch_list, return a nested
+     * list of those items for display when that attendee's merch is checked.
+     */
+    var createMerchList = function (message, merchItems) {
```
Comment (EliAndrewC) [inline uber/templates/registration/merch.html]: This is where we use the ``Attendee.merch_items`` list which the backend now passes to format a nicely formatted nested list.

File: uber/tests/models/test_merch.py [hunk 8]
```diff
@@ -164,24 +165,41 @@ def test_extra_merch(self):
...
-        assert 'RedShirt' in Attendee().merch and 'will be reported' in Attendee().merch
+        assert 'tshirt' in Attendee().merch and 'will be reported' in Attendee().merch
+
+    def test_paid_volunteer(self, paid_for_a_shirt, volunteer_event_shirt_eligible):
+        assert 'RedShirt' in Attendee(amount_extra=c.SHIRT_LEVEL).merch
+        assert '2nd tshirt' in Attendee(amount_extra=c.SHIRT_LEVEL).merch
 
     def test_volunteer_worked(self, volunteer_event_shirt_eligible, volunteer_event_shirt_earned):
-        assert 'RedShirt' == Attendee().merch
+        assert 'tshirt' in Attendee().merch
```
Comment (EliAndrewC) [inline uber/tests/models/test_merch.py]: I made one additional change to this implementation which is now reflected in our tests.  Previously we'd use whatever the textual description of the tshirt level was.  However, I noticed that in production that level is actually "Tshirt Pack".  So as things were implemented, we would have told the merch booth to give an entire tshirt pack to volunteers instead of just the shirt, which I don't believe is what we want.  (At least, that's never how it worked in the past back when kicking in at the tshirt level included a tshirt / button / stickers / ribbon.)

**Comment by EliAndrewC** [comment]
I believe I've now addressed all outstanding comments and suggestions.  I've also filed https://github.com/magfest/production-config/pull/353 to add the tier descriptions to the production config.  This PR and that one can be merged and deployed in any order.

**Comment by EliAndrewC** [comment]
> I trust you, so feel free to merge this

I'm headed to bed myself, and my rule of thumb is to never merge and deploy anything no matter how confident you are unless you have a spare hour to spend in case something goes wrong :)

I'll merge and deploy early tomorrow afternoon if you haven't had a chance to do it by then.  I should hopefully be on site by the time they start using this tomorrow (I believe the plan is to start handing out staff badges and swag at 6pm).
